### DISCUSSION TITLE
How to get each character coordinates of a sentence written? page.insert_text

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#4185] [Q&A]
page.insert_text

**Comment by JorjMcKie** [comment]
`page.insert_text(point, text)`, then `rect=page.seach_for(text)[0]`, then `blocks=page.get_text("rawdict",clip=rect)["blocks"]`.
Using the employed font, there also is `font.char_lengths(text, fontsize=11)` which gives a tuple of widths of the characters. Using `point`, you can compute the start of each character in `text`.
Using font size, ascender and descender, this also allows you to compute each char bbox directly

**Comment by nissansz** [comment]
Thank you. I can get letter info. Is it possible to change color and background color for the desired letters? How to change letter color /background color after I get below text box info.?

Character: 'H', BBox: (50, 89.0, 54.84253845306543, 100)
Character: 'e', BBox: (54.84253845306543, 89.0, 59.68507690613086, 100)
Character: 'l', BBox: (59.68507690613086, 89.0, 64.52761535919629, 100)
Character: 'l', BBox: (64.52761535919629, 89.0, 69.37015381226172, 100)
Character: 'o', BBox: (69.37015381226172, 89.0, 74.21269226532715, 100)
Character: ',', BBox: (74.21269226532715, 89.0, 79.05523071839258, 100)
Character: ' ', BBox: (79.05523071839258, 89.0, 83.89776917145801, 100)
Character: 'W', BBox: (83.89776917145801, 89.0, 88.74030762452344, 100)
Character: 'o', BBox: (88.74030762452344, 89.0, 93.58284607758887, 100)
Character: 'r', BBox: (93.58284607758887, 89.0, 98.4253845306543, 100)
Character: 'l', BBox: (98.4253845306543, 89.0, 103.26792298371973, 100)
Character: 'd', BBox: (103.26792298371973, 89.0, 108.11046143678516, 100)
Character: '!', BBox: (108.11046143678516, 89.0, 112.95299988985059, 100)

**Comment by nissansz** [comment]
![image](https://github.com/user-attachments/assets/8654dd59-108d-43ac-a002-16700609c444)
I tried to draw a rect for each letter with below code, which is very slow. Any fast method?
And the box position seems not accurate according to above screen capture, any better method? How to adjust below code?

```



import fitz  # PyMuPDF
import random

# 定义一个函数生成随机颜色
# Function to generate a random color
def random_color():
    return (random.random(), random.random(), random.random())  # Generates values between 0 and 1


# 创建新的 PDF 文档
doc = fitz.open()

# 添加一页到文档
page = doc.new_page()

# 要插入的文本和起始位置
text = "Hello, World!"
start_point = (50, 100)  # 起始位置 (x, y)

# 插入文本
page.insert_text(start_point, text, fontsize=11)

# 搜索文本以获取其矩形
rect = page.search_for(text)[0]  # 获取找到的第一个矩形

# 获取指定矩形内的文本块
blocks = page.get_text("rawdict", clip=rect)["blocks"]

# 假设文本存储在第一个块中
char_positions = []
if blocks:
    for block in blocks:
        if "lines" in block:  # 确保块中包含行
            # 获取字体信息
            span = block["lines"][0]["spans"][0]
            fontname = span["font"]  # 字体名称
            fontsize = span["size"]  # 字体大小

            # 初始化 x 坐标
            x = start_point[0]
            y = start_point[1]

            # 遍历每个字符，计算它们的宽度并记录位置
            total_width = fitz.get_text_length(text, fontsize=fontsize, fontname=fontname)
            char_widths = [total_width / len(text)] * len(text)  # 假设字符宽度相同

            for i, char in enumerate(text):
                bbox = (x, y - fontsize, x + char_widths[i], y)  # 计算边界框
                char_positions.append({
                    'char': char,
                    'bbox': bbox  # 每个字符的边界框
                })
                x += char_widths[i]  # 移动 x 到下一个字符的起始位置
                # page.draw_rect(bbox, color=None, fill=fitz.pdfcolor["yellow"], fill_opacity=0.5, overlay=True)
                page.draw_rect(bbox, color=None, fill=random_color(), fill_opacity=0.5, overlay=True)

# 打印字符位置
for position in char_positions:
    print(f"Character: '{position['char']}', BBox: {position['bbox']}")

# 保存文档
doc.save("output.pdf")
doc.close()
```

**Comment by nissansz** [comment]
When I try to use a local font, below error occurs. How to solve it?

    total_width = fitz.get_text_length(text, fontsize=11, fontname='F0')  # 使用字体文件名
  File "D:\Python3.8.8\lib\site-packages\pymupdf\__init__.py", line 18203, in get_text_length
    raise ValueError("Font '%s' is unsupported" % fontname)
ValueError: Font 'f0' is unsupported
'''

# 指定字体文件路径
font_path = r'F:\fonts\SarasaTermSCNerd.ttc'  # 下载字体文件simhei.ttf
#
# font_path = "path/to/your/font.ttf"  # 替换为你的字体文件路径
font = fitz.Font(fontfile=font_path)  # pymupdf加载字体， choose a font with small caps

page.insert_font(fontname="F0", fontbuffer=font.buffer)

# # 加载自定义字体
# font = fitz.Font(font_path)

# 插入文本
# page.insert_text(start_point, text, fontsize=11, font=font)  # 使用加载的自定义字体
page.insert_text(start_point, text, fontsize=11, fontname="F0")  # 使用加载的自定义字体

# 搜索文本以获取其矩形
rect = page.search_for(text)[0]  # 获取找到的第一个矩形


# 获取文本的总宽度
total_width = fitz.get_text_length(text, fontsize=11, fontname='F0')  # 使用字体文件名

'''

**Comment by JorjMcKie** [comment]
You can **_never_** change the color of text that has been written. Background color is possible.
You should **_either_** extract via rawdict - in which case you get the bboxes for free - **_or_** use the alternative computing the bbox yourself, but not both.
The rawdict way can be simplified:
```python
blocks = page.get_text("rawdict", clip=clip, flags=pymupdf.TEXTFLAGS_TEXT)["blocks"]
# list of all characters
chars = [c for b in blocks for l in b["lines"] for s in l["spans"] for c in s["chars"]]
for c in chars:
    color = random ... whatever
    page.draw_rect(c["bbox"], fill_color=color, ...)
```
There is no faster way.

As for using your own font:
I think you've made that same error before:
(1) insert the font in the page giving it an appropriate "fontname" like "F0". **_This is not the filename!_**
(2) When inserting text use `fontname="F0"`. There is no parameter "font" available in this method!

Finally, please do everybody a favor and make your code examples readable by using markdown code blocks. These are groups of lines starting and ending with 3 backticks. You can use this symbol if you need help: 
![image](https://github.com/user-attachments/assets/7e6b0a59-15f9-47d7-9e48-2bd228a43d92)

**Reply by JorjMcKie** [reply]
The method `pymupdf.get_text_length` is only supported for Base14 fonts - please consult the documentation.
For other fonts use the `Font` object and method `Font.text_length`.

**Reply by nissansz** [reply]
![image](https://github.com/user-attachments/assets/2071cac7-66af-4524-9de7-4cff571e5429)
Cannot see any effect using / or ///.

**Reply by JorjMcKie** [reply]
Give me a break! This is what must be used (backtick!):
![image](https://github.com/user-attachments/assets/0c68cdf1-3263-4533-9f5c-dce38b3e5c33)

**Reply by nissansz** [reply]
It works. Due to vpn slow connection, pop up window appears slow, so cannot see the options originally.
``` ```

**Reply by nissansz** [reply]
How to use local font for below code block?
font_path = r'F:\fonts\SarasaTermSCNerd.ttc'  # 下载字体文件simhei.ttf

```
# 创建新的 PDF 文档
doc = fitz.open()

# 添加一页到文档
page = doc.new_page()

# 要插入的文本和起始位置
text = "Hello, World!"
start_point = (50, 100)  # 起始位置 (x, y)

# 插入文本
page.insert_text(start_point, text, fontsize=11)

# 搜索文本以获取其矩形
rect = page.search_for(text)[0]  # 获取找到的第一个矩形

blocks = page.get_text("rawdict", clip=rect, flags=pymupdf.TEXTFLAGS_TEXT)["blocks"]
# list of all characters
chars = [c for b in blocks for l in b["lines"] for s in l["spans"] for c in s["chars"]]
for c in chars:
    # color = random ... whatever
    # page.draw_rect(c["bbox"], fill_color=color, ...)
    page.draw_rect(c["bbox"], color=None, fill=random_color(), fill_opacity=0.5, overlay=True)
#
```

**Reply by JorjMcKie** [reply]
Uff - finally!
```python
import pymupdf

doc = pymupdf.open()
RED = (1, 0, 0)
BLUE = (0, 0, 1)
GREEN = (0, 1, 0)
colors = (RED, GREEN, BLUE)
page = doc.new_page()
fontfile = "c:/windows/fonts/simsunb.ttf"  # example only
text = "Hello, World!"
start_point = (50, 100)  # 起始位置 (x, y)
page.insert_font(fontname="F0", fontfile=fontfile)
page.insert_text(start_point, text, fontsize=11, fontname="F0")

rect = page.search_for(text)[0]

blocks = page.get_text("rawdict", clip=rect, flags=pymupdf.TEXTFLAGS_TEXT)["blocks"]
# list of all characters
chars = [c for b in blocks for l in b["lines"] for s in l["spans"] for c in s["chars"]]
for i, c in enumerate(chars):
    page.draw_rect(
        c["bbox"], color=None, fill=colors[i % 3], fill_opacity=0.5, overlay=True
    )
doc.ez_save("x.pdf")
```
Gives this:
![image](https://github.com/user-attachments/assets/bedb7336-320a-4798-b59a-cab6f94df12d)

**Reply by nissansz** [reply]
Great! It works. 
I found a script on documentation below. I wonder whether it is possible to apply font size to below and calculate coordinates before writing text, to choose font color for each letter desired, and finally apply morph for a line wiht different font colors for each letter desired.

font = pymupdf.Font("helv")
text = "PyMuPDF"
font.text_length(text)
50.115999937057495
pymupdf.get_text_length(text, fontname="helv")
50.115999937057495
sum(font.char_lengths(text))
50.115999937057495
pprint(font.char_lengths(text))
(7.336999952793121,  # P
5.5,                 # y
9.163000047206879,   # M
6.115999937057495,   # u
7.336999952793121,   # P
7.942000031471252,   # D
6.721000015735626)   # F

**Reply by JorjMcKie** [reply]
Use the font's ascender / descender, the font size and the insertion point. Then compute the text length as shown. This is the bbox width.
The insertion point's x value is x0 of the bbox. x1 is x0 + width.
y1 of the bbox is insertion point `y - font.descender*fontsize`, y0 is `y - font.ascender*fontsize`.

**Reply by nissansz** [reply]
Used your previous code, effect is much bettern than before for 3 text_writers.

![image](https://github.com/user-attachments/assets/6ceee4e1-998b-455e-bd79-2081dd255d10)

**Comment by nissansz** [comment]
Another question is for below code. If there are 100 texts on one page, text may be either same or different.
Any other option to locate the text written just now instead of parsing all text written before? 

rect = page.search_for(text)[0]

**Reply by JorjMcKie** [reply]
No

**Comment by nissansz** [comment]
Is space not included in blocks?

source chars: ['賃', '金', '規', '程', '第', ' ', '4']

blocks: missing one space
[{'origin': (358.9205017089844, 515.3494873046875), 'bbox': (358.9205017089844, 509.5594787597656, 368.8268737792969, 517.0595092773438), 'c': '賃'}, {'origin': (368.8268737792969, 515.3494873046875), 'bbox': (368.8268737792969, 509.5594787597656, 378.7332458496094, 517.0595092773438), 'c': '金'}, {'origin': (378.7332458496094, 515.3494873046875), 'bbox': (378.7332458496094, 509.5594787597656, 388.6396179199219, 517.0595092773438), 'c': '規'}, {'origin': (388.6396179199219, 515.3494873046875), 'bbox': (388.6396179199219, 509.5594787597656, 398.5459899902344, 517.0595092773438), 'c': '程'}, {'origin': (398.5459899902344, 515.3494873046875), 'bbox': (398.5459899902344, 509.5594787597656, 408.4523620605469, 517.0595092773438), 'c': '第'}, {'origin': (413.4055480957031, 515.3494873046875), 'bbox': (413.4055480957031, 509.5594787597656, 418.3587341308594, 517.0595092773438), 'c': '4'}]

**Comment by nissansz** [comment]
Ending dot box may be wrong. And there may be some additional spaces at the end of line.

![image](https://github.com/user-attachments/assets/c84f72fc-d84e-4bc1-ac4e-af3df81f4498)

[x.pdf](https://github.com/user-attachments/files/18306116/x.pdf)
